agent/consul/stream/event_publisher.go

```diff
@@ -197,11 +266,17 @@ func (e *EventPublisher) Subscribe(req *SubscribeRequest) (*Subscription, error)
 		Payload: newSnapshotToFollow{},
 	}})
 	result.buffer.AppendItem(snapFromCache.First)
-	return e.subscriptions.add(req, result.First), nil
+	return e.subscriptions.add(req, result.First, freeBuf), nil
 }
 
-func (s *subscriptions) add(req *SubscribeRequest, head *bufferItem) *Subscription {
-	sub := newSubscription(*req, head, s.unsubscribe(req))
+func (s *subscriptions) add(req *SubscribeRequest, head *bufferItem, freeBuf func()) *Subscription {
+	// We wrap freeBuf in a sync.Once as it's expected that Subscription.unsub is
+	// idempotent, but freeBuf decrements the reference counter on every call.
+	var once sync.Once
+	sub := newSubscription(*req, head, func() {
+		s.unsubscribe(req)
+		once.Do(freeBuf)
```

Comment:
I stopped to think for a minute about whether it matters that these two calls each take the lock separately. It means they are not atomic. i.e. there is a point in time where the reference count of subscriptions is higher that the sum of all the subscriptions in the subscription maps.

Assuming the code can't panic or exit in some way before the freeBuf call, I _think_ this is OK eventually, it just might mean we defer garbage collecting for a while after the subscription has been removed because we are waiting for the lock a second time.

 The sync.Once ensures it can't happen multiple times though so it should still always be correct eventually.

An alternative approach might be to document that `freeBuf` is called with the lock held, and then not take it inside the closure above, and then pass it in to `unsubscribe()` to be executed while the lock is held. That has a marginal performance gain under contention for not needing to take and release the lock twice in a single unsubscribe but i don't think that's important. I think this would have the other minor benefit of not needing the sync.Once. That is saving an allocation but it's not in a critical path so the main benefit would be simpler code.

What do you think?  I don't _think_ it matter from a correctness perspective though it's a little harder to reason about this way.

Reply:
> these two calls each take the lock separately

I think they take separate locks? (`EventPublisher.lock` and `subscriptions.lock`). The point about atomicity still stands, though!

> I think this would have the other minor benefit of not needing the sync.Once.

To be honest, I've sort of taken the assertion that `unsub` needs to be idempotent (from its doc comment) as a given, and haven't taken the time to dig into why that's the case. Wouldn't we still need some way to prevent decrementing the reference counter twice?

> What do you think? I don't think it matter from a correctness perspective though it's a little harder to reason about this way.

Agreed! The dependencies here are a bit hard to grok. That said, without merging `subscriptions` into `EventPublisher`, I can't think of an obvious way to share the lock that isn't equally complex (e.g. passing `freeBuf` to `unsubscribe` as you mentioned).

Reply:
> I think they take separate locks? (EventPublisher.lock and subscriptions.lock). The point about atomicity still stands, though!

Ah yes, my mistake. That makes it a bit more complicated 🤔 .

We currently rely on subscription lock being taken while publisher lock is held during `EventPublisher.Subscribe` so to make it atomic we'd need to ensure that `unsubscribe` first took `EventPublisher.Lock` before taking the `subscribers.lock` or it would deadlock. This has a slight potentially increase for contention over this implementation - unsubscribing would hold the `EventPublisher.lock` slightly longer (blocking Publish) since it would also include the subscription updates, but we already hold both for `Subscribe` which is probably more expensive anyway so I don't expect that to be a real issue.

So I think it would still be possible to do what I suggested, just with the addition of taking both locks in `unsubscribe` in the right order... but I'm also fine with leaving it as I can't think of a way it could actually stay in an invalid state right and we don't assume the ref count is perfectly consistent right now. Maybe we should document that the `freeBuf` call is not called under any locks and so its effects are not atomic/consistent with other subscription state?

> Wouldn't we still need some way to prevent decrementing the reference counter twice?

My assumption was that you'd get that for free thanks to the early return in `unsubscribe` if the sub state doesn't exist. If we removed all state atomically that would also mean the buffer state had been cleared too do nothing further needed. But that was under the assumption that it was the same lock being held in `unsubscribe` and `freeBuf` which was wrong.

So yeah I don't think there is anything substantially simpler here - the reason subscriptions have a separate lock anyway is to avoid blocking the publish mechanism for subscription activity.